File: backend/app/models/firm.py
"""SQLAlchemy models — firm domain."""
from __future__ import annotations

import enum
import uuid
from decimal import Decimal
from datetime import date, datetime, time

from sqlalchemy import (
    BigInteger,
    Boolean,
    Date,
    DateTime,
    Float,
    LargeBinary,
    Time,
    Enum,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    SmallInteger,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base

class LetterheadStyle(str, enum.Enum):
    """Letter compose: physical pre-printed stock vs digital header/footer template."""

    preprinted = "preprinted"
    digital = "digital"

class PrecedentKind(str, enum.Enum):
    letter = "letter"
    email = "email"
    document = "document"

class PrecedentCategory(Base):
    __tablename__ = "precedent_category"
    __table_args__ = (UniqueConstraint("matter_sub_type_id", "name", name="uq_precedent_category_sub_name"),)

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    matter_sub_type_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("matter_sub_type.id", ondelete="RESTRICT"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

class FirmSettings(Base):
    """Singleton firm-wide configuration (id must always be 1)."""

    __tablename__ = "firm_settings"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, default=1)
    trading_name: Mapped[str] = mapped_column(String(300), nullable=False, default="")
    registered_company_name: Mapped[str | None] = mapped_column(String(400), nullable=True)
    addr_line1: Mapped[str | None] = mapped_column(String(300), nullable=True)
    addr_line2: Mapped[str | None] = mapped_column(String(300), nullable=True)
    town_city: Mapped[str | None] = mapped_column(String(200), nullable=True)
    county: Mapped[str | None] = mapped_column(String(150), nullable=True)
    postcode: Mapped[str | None] = mapped_column(String(50), nullable=True)
    letterhead_style: Mapped[LetterheadStyle] = mapped_column(
        Enum(LetterheadStyle, name="letterhead_style"),
        nullable=False,
        default=LetterheadStyle.preprinted,
    )
    letterhead_file_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("file.id", ondelete="SET NULL"), nullable=True
    )
    quote_letterhead_style: Mapped[LetterheadStyle] = mapped_column(
        Enum(LetterheadStyle, name="letterhead_style"),
        nullable=False,
        default=LetterheadStyle.preprinted,
    )
    quote_letterhead_file_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("file.id", ondelete="SET NULL"), nullable=True
    )
    portal_logo_file_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("file.id", ondelete="SET NULL"), nullable=True
    )
    # Client portal page background only (#RRGGBB). Null = product default chrome.
    portal_background_color: Mapped[str | None] = mapped_column(String(7), nullable=True)
    default_signature_file_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("file.id", ondelete="SET NULL"), nullable=True
    )
    default_signature_scale: Mapped[int] = mapped_column(Integer, nullable=False, default=7)
    invoice_template_file_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("file.id", ondelete="SET NULL"), nullable=True
    )
    mandate_two_factor: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    mandate_password_rotation: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    password_rotation_days: Mapped[int | None] = mapped_column(Integer, nullable=True)
    storage_limit_bytes: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    client_bank_account_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    client_bank_sort_code: Mapped[str | None] = mapped_column(String(16), nullable=True)
    client_bank_account_number_last4: Mapped[str | None] = mapped_column(String(4), nullable=True)
    client_bank_account_number: Mapped[str | None] = mapped_column(String(20), nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

class ReconciliationStatus(str, enum.Enum):
    draft = "draft"
    approved = "approved"

class ClientAccountReconciliation(Base):
    """Month-end client account reconciliation snapshot (one row per period end date)."""

    __tablename__ = "client_account_reconciliation"
    __table_args__ = (UniqueConstraint("period_end_date", name="uq_client_account_reconciliation_period_end"),)

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    period_end_date: Mapped[date] = mapped_column(Date, nullable=False)
    ledger_client_total_pence: Mapped[int] = mapped_column(Integer, nullable=False)
    ledger_office_total_pence: Mapped[int] = mapped_column(Integer, nullable=False)
    bank_statement_balance_pence: Mapped[int] = mapped_column(Integer, nullable=False)
    difference_pence: Mapped[int] = mapped_column(Integer, nullable=False)
    prepared_by_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("user.id", ondelete="RESTRICT"), nullable=False
    )
    prepared_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    approved_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("user.id", ondelete="SET NULL"), nullable=True
    )
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[ReconciliationStatus] = mapped_column(
        Enum(ReconciliationStatus, name="reconciliation_status"),
        nullable=False,
        default=ReconciliationStatus.draft,
    )

class MergeCodeCatalog(Base):
    """Editable descriptions for precedent merge tokens; keys mirror ``docx_util.PRECEDENT_CODES``."""

    __tablename__ = "merge_code_catalog"

    code: Mapped[str] = mapped_column(String(160), primary_key=True)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

class Precedent(Base):
    __tablename__ = "precedent"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(300), nullable=False)
    reference: Mapped[str] = mapped_column(String(200), nullable=False)
    kind: Mapped[PrecedentKind] = mapped_column(Enum(PrecedentKind, name="precedent_kind"), nullable=False)
    file_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("file.id", ondelete="CASCADE"), nullable=False)
    # Scope: (NULL,NULL,NULL) = all cases; (H,NULL,NULL) = all sub-types under head H; (H,S,NULL) = all categories under sub S; (H,S,C) = one category.
    matter_head_type_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("matter_head_type.id", ondelete="RESTRICT"), nullable=True, index=True
    )
    matter_sub_type_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("matter_sub_type.id", ondelete="RESTRICT"), nullable=True, index=True
    )
    category_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("precedent_category.id", ondelete="RESTRICT"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

class EmailIntegrationSettings(Base):
    """Singleton row (id=1): mailto vs Microsoft Graph; optional Entra app credentials (secret encrypted)."""

    __tablename__ = "email_integration_settings"

    id: Mapped[int] = mapped_column(SmallInteger, primary_key=True)
    integration_mode: Mapped[str] = mapped_column(String(32), nullable=False, default="microsoft_graph")
    graph_tenant_id: Mapped[str | None] = mapped_column(Text, nullable=True)
    graph_client_id: Mapped[str | None] = mapped_column(Text, nullable=True)
    graph_client_secret_enc: Mapped[str | None] = mapped_column(Text, nullable=True)
    outlook_web_mail_base: Mapped[str | None] = mapped_column(Text, nullable=True)
    alerts_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    alert_transport: Mapped[str] = mapped_column(String(16), nullable=False, default="auto")
    graph_send_mailbox: Mapped[str | None] = mapped_column(String(320), nullable=True)
    graph_send_from_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

class SmtpNotificationSettings(Base):
    """Singleton (id=1): outbound SMTP for calendar (and future) e-mail alerts."""

    __tablename__ = "smtp_notification_settings"

    id: Mapped[int] = mapped_column(SmallInteger, primary_key=True)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    host: Mapped[str | None] = mapped_column(String(300), nullable=True)
    port: Mapped[int] = mapped_column(Integer, nullable=False, default=587)
    use_tls: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    username: Mapped[str | None] = mapped_column(String(320), nullable=True)
    password_enc: Mapped[str | None] = mapped_column(Text, nullable=True)
    from_email: Mapped[str | None] = mapped_column(String(320), nullable=True)
    from_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

File: frontend/src/types/firm.ts
export type LetterheadStyle = 'preprinted' | 'digital'

export type MergeCodeCatalogOut = {
  code: string
  description: string
  sort_order: number
}

export type MergeCodeCatalogImportResult = {
  updated: number
  skipped_unknown: number
}

export type FirmSettingsOut = {
  id: number
  trading_name: string
  registered_company_name?: string | null
  addr_line1?: string | null
  addr_line2?: string | null
  town_city?: string | null
  county?: string | null
  postcode?: string | null
  letterhead_style: LetterheadStyle
  letterhead_original_filename?: string | null
  quote_letterhead_style?: LetterheadStyle
  quote_letterhead_original_filename?: string | null
  portal_logo_configured?: boolean
  portal_logo_original_filename?: string | null
  /** `#RRGGBB` solid portal page background; null = product default. */
  portal_background_color?: string | null
  default_signature_configured?: boolean
  default_signature_original_filename?: string | null
  default_signature_scale?: number
  mandate_two_factor?: boolean
  mandate_password_rotation?: boolean
  password_rotation_days?: number | null
  client_bank_account_name?: string | null
  client_bank_sort_code?: string | null
  client_bank_account_number_last4?: string | null
  client_bank_account_number?: string | null
}

export type ClientAccountReconciliationOut = {
  id: string
  period_end_date: string
  ledger_client_total_pence: number
  ledger_office_total_pence: number
  bank_statement_balance_pence: number
  difference_pence: number
  prepared_by_user_id: string
  prepared_by_name?: string | null
  prepared_at: string
  approved_by_user_id?: string | null
  approved_by_name?: string | null
  approved_at?: string | null
  notes?: string | null
  status: 'draft' | 'approved'
}

export type ReconciliationPreviewOut = {
  ledger_client_total_pence: number
  ledger_office_total_pence: number
}

export type AccountantPackSectionOut = {
  key: string
  label: string
  included: boolean
  row_count?: number | null
  note?: string | null
}

export type AccountantPackPreviewOut = {
  period_end_date: string
  activity_date_from: string
  activity_date_to: string
  fee_earner_count: number
  reconcile_doc_available: boolean
  sections: AccountantPackSectionOut[]
}

/** Registered passkey row from GET /auth/webauthn/credentials */

export type PrecedentCategoryOut = {
  id: string
  matter_sub_type_id: string
  name: string
  sort_order: number
  created_at: string
  updated_at: string
}

export type PrecedentCategoryFlatOut = PrecedentCategoryOut & {
  matter_sub_type_name: string
}

/** Form/API token for Global scope (must match backend GLOBAL_SCOPE). */

export const GLOBAL_PRECEDENT_SCOPE = '__GLOBAL__'

export type PrecedentOut = {
  id: string
  name: string
  reference: string
  kind: 'letter' | 'email' | 'document'
  original_filename: string
  mime_type: string
  category_id?: string | null
  matter_head_type_id?: string | null
  matter_sub_type_id?: string | null
  category_name?: string | null
  matter_head_type_name?: string | null
  matter_sub_type_name?: string | null
  scope_summary?: string
  created_at: string
}

export type EmailIntegrationSettingsOut = {
  integration_mode: 'mailto' | 'microsoft_graph'
  graph_tenant_id: string | null
  graph_client_id: string | null
  graph_client_secret_configured: boolean
  outlook_web_mail_base: string | null
  alerts_enabled: boolean
  alert_transport: 'auto' | 'graph' | 'smtp'
  graph_send_mailbox: string | null
  graph_send_from_name: string | null
  graph_alert_ready: boolean
  smtp_alert_ready: boolean
  effective_alert_transport: 'graph' | 'smtp' | null
}

export type CaseSourceOut = {
  id: string
  name: string
  sort_order: number
  is_system: boolean
}

// Sub-menu Events (admin template + case rows)

export type AdminDeployUpdateCheckOut = {
  github_repo_configured: boolean
  deploy_trigger_configured: boolean
  compose_update_enabled: boolean
  compose_git_reset_enabled?: boolean
  compose_git_ref?: string
  prompt_enabled: boolean
  current_commit: string
  current_commit_short: string
  remote_ref: string
  remote_commit: string
  remote_commit_short: string
  update_available: boolean
  build_commit_unknown: boolean
  compare_html_url?: string | null
  latest_release_tag?: string | null
  latest_release_name?: string | null
  latest_release_body?: string | null
  commit_messages: string[]
  note?: string | null
}

export type AdminStorageCategoryOut = {
  category: string
  label: string
  bytes_used: number
  file_count: number
}

export type AdminStorageDeploymentComponentOut = {
  key: string
  label: string
  bytes_used: number
  detected: boolean
}

export type AdminStorageOut = {
  tracked_total_bytes: number
  files_on_disk_bytes: number
  compose_mount_bytes: number
  application_checkout_bytes: number
  database_bytes: number | null
  database_logical_bytes: number | null
  calendars_bytes: number | null
  deployment_total_bytes: number
  docker_detected: boolean
  docker_images_bytes: number
  docker_container_writable_bytes: number
  docker_dangling_images_bytes: number
  docker_build_cache_bytes: number | null
  deployment_active_bytes: number
  deployment_artifacts_bytes: number
  measurement_note: string | null
  deployment_components: AdminStorageDeploymentComponentOut[]
  categories: AdminStorageCategoryOut[]
  storage_limit_bytes: number | null
  files_root: string
  host_disk_detected: boolean
  host_disk_total_bytes: number | null
  host_disk_used_bytes: number | null
  host_disk_free_bytes: number | null
}

export type AdminDeployTriggerOut = {
  ok: boolean
  message: string
  async_mode?: boolean
  job_id?: string | null
}

export type AdminDeployComposeJobOut = {
  status: 'idle' | 'running' | 'succeeded' | 'failed'
  job_id?: string | null
  started_at?: string | null
  finished_at?: string | null
  message?: string | null
  error_detail?: string | null
  log_excerpt?: string | null
  journal_lines?: string[]
  progress_phase?: 'git' | 'build' | 'up' | null
  elapsed_seconds?: number | null
}

export type SmtpNotificationSettingsOut = {
  enabled: boolean
  host: string | null
  port: number
  use_tls: boolean
  username: string | null
  password_configured: boolean
  from_email: string | null
  from_name: string | null
}
